File: python/organisms/animals/turtle.py
from gui.logger import Logger
from organisms.animals.animal import Animal
from organisms.species import Species
from utils.helpers import Helpers


class Turtle(Animal):
    def __init__(self, pos):
        super().__init__(pos)
        self.strength = 2
        self.initiative = 1
        self.species = Species.TURTLE

    def get_new_point(self, grid):
        if Helpers.is_lucky(75):
            return None
        return super().get_new_point(grid)

    def collision(self, other, attacking, grid):
        if other.strength < 5 and not attacking:
            Logger.log('Turtle reflected attack from: ' + other.species.name)
            return True
        return False

File: java/pl/edu/pg/mygame/organisms/animals/Turtle.java
package pl.edu.pg.mygame.organisms.animals;

import pl.edu.pg.mygame.Grid;
import pl.edu.pg.mygame.World;
import pl.edu.pg.mygame.organisms.Organism;
import pl.edu.pg.mygame.organisms.Species;

import java.awt.*;

import static pl.edu.pg.mygame.utils.HelperFunctions.isLucky;

public class Turtle extends Animal {

    public Turtle(Point p) {
        super(p);
        this.species = Species.TURTLE;
        this.initiative = 1;
        this.strength = 2;
    }

    @Override
    protected Point getNewPoint(Grid grid) {
        if (isLucky(75))
            return null;
        return super.getNewPoint(grid);
    }

    @Override
    public boolean collision(Organism other, boolean attacking, Grid grid) {
        if (other.getStrength() < 5 && !attacking) {
            World.message_panel.showMessage("Turtle reflected attack from: " + other.getSpecies() + "\n");
            return true;
        }
        return false;
    }

}
